### ISSUE TITLE
TypeError: Cannot assign to read only property 'constructor'

### ISSUE THREAD

**Issue by blakeembrey** [Microsoft/TypeScript#6887]
With the TypeScript 1.8 change of `'use strict'`, the emitted JavaScript (specifically `__extends` helper) can be invalid.

**TypeScript:**

```ts
import makeError = require('make-error')

export class SubError extends makeError.BaseError {}
```

**Emitted JavaScript:**

```js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var makeError = require('make-error');
var SubError = (function (_super) {
    __extends(SubError, _super);
    function SubError() {
        _super.apply(this, arguments);
    }
    return SubError;
}(makeError.BaseError));
exports.SubError = SubError;
```

**Error:**

```
/Users/blakeembrey/Projects/tmp/test.js:4
    function __() { this.constructor = d; }
                                     ^

TypeError: Cannot assign to read only property 'constructor' of Error
    at BaseError.__ (/Users/blakeembrey/Projects/tmp/test.js:4:38)
    at __extends (/Users/blakeembrey/Projects/tmp/test.js:5:80)
    at /Users/blakeembrey/Projects/tmp/test.js:9:5
```

**Comment by blakeembrey** [comment]
@RyanCavanaugh Changing this line to define the property instead solves this issue: `Object.defineProperty(this, 'constructor', { value: d })`. I can make a PR, my only question is whether this should be detected at runtime (E.g. check for `Object.defineProperty` existence before using) or if it should emit this syntax only when emitting to ES5?

**Comment by RyanCavanaugh** [comment]
Is it the case that any runtime which lacks `Object.defineProperty` will allow the `constructor` assignment? @bterlson ?

**Comment by blakeembrey** [comment]
Not definitive, but you can't have gotten into this case unless you used `defineProperty` to begin with (see https://github.com/julien-f/js-make-error/blob/master/index.js#L76-L82). I don't know what the chance of an ES3 environment is that has a native read only properly **and** support `'use strict'` (won't ES3 envs ignore it, I can't recall the exact timeline of introduction here).

**Comment by rbuckton** [comment]
@blakeembrey The reason this fails is due to the fact that the `make-error` project explicitly marks `constructor` as non-writable, non-configurable. 

@RyanCavanaugh Any ES3 runtime will not support defineProperty. IIRC, IE8 supports defineProperty, but only on DOM nodes.

**Comment by rbuckton** [comment]
We would have the same issue if someone were to write the following:

```ts
class A {}
Object.freeze(A.prototype);
class B extends A { }
```

Even with this @blakeembrey's proposed change, we would still error in the following case:

```ts
class A {
  method() {}
}
Object.freeze(A.prototype);
class B extends A { 
  method() {}
}
```

**Comment by blakeembrey** [comment]
@rbuckton I know, mentioned above. It's marked on the prototype only though, it can be set on the new prototype object being created by using `defineProperty`. 

I'll have to check that new example, I'm not sure it's accurate. I think it'll only stop modifications to that prototype, but not the new object. But that's speculation, give me 5 minutes.

**Comment by blakeembrey** [comment]
@rbuckton Yeah, I just checked - it works there too. Freeze will only freeze the actual object, it has no impact on new objects that extend from a frozen object.

**Comment by rbuckton** [comment]
@blakeembrey I just tested the following in node v5:

```js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { Object.defineProperty(this, "constructor", { value: d, configurable: true, writable: true; }); }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var A = (function() { 
  function A() { }
  A.prototype.method = function() {} 
  return A;
})();

Object.freeze(A.prototype);

var B = (function (_super) {
  __extends(B, _super);
  function B() {}
  B.prototype.method = function() {} // TypeError: Cannot assign to read only property 'method' of #<B>
  return B;
})(A);
```

**Comment by blakeembrey** [comment]
@rbuckton Yes, assignment will break - I thought we were just talking about the constructor. It's because the shadowing breaks it. If you want to fix it all properly, everything should use `defineProperty` (including setting the prototype methods). 

Is there a way in TypeScript to denote that something is not configuration (I saw the `readonly` flag, but haven't looked if it would solve it). If such a thing existed, it could be used to conditionally change between the assignment vs `defineProperty` ways of adding methods.

**Comment by rbuckton** [comment]
@blakeembrey I don't think we plan to alter the emit based on whether someone has added a `readonly` modifier.

**Comment by blakeembrey** [comment]
Would there be a consideration to always emit using `defineProperty` then? This is blocking an update to TS 1.8, but if there's no resolution I guess I'll have to resort to asking the author to change the configuration.

**Comment by rbuckton** [comment]
@blakeembrey I discussed offline with @mhegazy and I don't think this is something that we will fix. However, we recently added a `--noImplicitUseStrict` compiler option which would allow you to revert to our previous behavior.

**Comment by blakeembrey** [comment]
That's unfortunate. I'd prefer to have `'use strict'` by default. Can you elaborate on the reason why not? It seems like something that should be handled, even Babel does it by default but I'm sure with the type information it can be clever when to apply it.

**Comment by rbuckton** [comment]
There are a number of related issues that we have, up to this point, opted not to address as they are often extreme corner cases. This not only includes things like using `Object.defineProperty` for `constructor`, but also for copying static members, and defining methods on classes. While these seem to be easily resolved issues, there are various constraints that make the cost associated with these changes much too high with respect to the resulting value to the user base. 

While we may address many of these concerns in the long term, we would more than likely need to balance these issues against our current priorities.

**Comment by RyanCavanaugh** [comment]
With increasing ES6 runtime adoption we're probably not going to complicate this area any further unless there's a pressing need with no workaround.

**Comment by zbjornson** [comment]
Yet another way this situation arises is when the base class is written with

```js
MyClass.prototype = Object.create(Stream.prototype, {constructor: {value: MyClass}});
```

(e.g. https://github.com/isaacs/sax-js/blob/master/lib/sax.js#L222-L226).

If it helps anyone else, adding this sort of stub seems to be a workaround:

```js
sax.SAXStream.prototype = Object.create(sax.SAXStream.prototype, {
	constructor: {
		value: sax.SAXStream,
		writable: true
	}
});
```

**Comment by jpea-imvu** [comment]
@blakeembrey @RyanCavanaugh @rbuckton You guys support the `"module":"es2015"` settings, and ES6 modules are strict mode by default, so it seems you guys should by default support the use of `"use strict"` for at least `"module":"es2015"` mode.

For example, `ts-loader` in passes modules to Webpack in ES2015 format when `"module":"es2015"`, then Webpack does the right thing and automatically adds `"use strict"` to the final module inside the bundle, which causes this error.

Example typescript:

```ts
import {UiContext} from '@UiCore'

function main() {

    // Yay, we can use ES6 classes!
    class App extends UiContext {
        uiApplication: string
        uiContextName: string

        initialize(options:Object) {
            this.uiApplication = 'sticker_creator'
            this.uiContextName = 'app'

            super.initialize(options)

            this.addToDOM()
        }
    }
}

main()
```

TypeScript compiles it to

```js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import { UiContext } from '@UiCore';
function main() {
    // Yay, we can use ES6 classes!
    var App = (function (_super) {
        __extends(App, _super);
        function App() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        App.prototype.initialize = function (options) {
            this.uiApplication = 'sticker_creator';
            this.uiContextName = 'app';
            _super.prototype.initialize.call(this, options);
            this.addToDOM();
        };
        return App;
    }(UiContext));
}
main();
```

and then Webpack wraps it and includes `"use strict"`:

```js
/***/ 185:
/***/ (function(module, exports, __webpack_require__) {

"use strict";

var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var _UiCore_1 = __webpack_require__(56);
function main() {
    // Yay, we can use ES6 classes!
    var App = (function (_super) {
        __extends(App, _super);
        function App() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        App.prototype.initialize = function (options) {
            this.uiApplication = 'sticker_creator';
            this.uiContextName = 'app';
            _super.prototype.initialize.call(this, options);
            this.addToDOM();
        };
        return App;
    }(_UiCore_1.UiContext));
}
main();


/***/ }),
```

**Comment by trusktr** [comment]
@blakeembrey @RyanCavanaugh @rbuckton You guys support the `"module":"es2015"` settings, and ES6 modules are strict mode by default, so it seems you guys should by default support the use of `"use strict"` for at least `"module":"es2015"` mode.

For example, `ts-loader` in passes modules to Webpack in ES2015 format when `"module":"es2015"`, then Webpack does the right thing and automatically adds `"use strict"` to the final module inside the bundle, which causes this error.

Example typescript:

```ts
import {UiContext} from '@UiCore'

function main() {

    // Yay, we can use ES6 classes!
    class App extends UiContext {
        uiApplication: string
        uiContextName: string

        initialize(options:Object) {
            this.uiApplication = 'sticker_creator'
            this.uiContextName = 'app'

            super.initialize(options)

            this.addToDOM()
        }
    }
}

main()
```

TypeScript compiles it to

```js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import { UiContext } from '@UiCore';
function main() {
    // Yay, we can use ES6 classes!
    var App = (function (_super) {
        __extends(App, _super);
        function App() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        App.prototype.initialize = function (options) {
            this.uiApplication = 'sticker_creator';
            this.uiContextName = 'app';
            _super.prototype.initialize.call(this, options);
            this.addToDOM();
        };
        return App;
    }(UiContext));
}
main();
```

and then Webpack wraps it and includes `"use strict"`:

```js
/***/ 185:
/***/ (function(module, exports, __webpack_require__) {

"use strict";

var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var _UiCore_1 = __webpack_require__(56);
function main() {
    // Yay, we can use ES6 classes!
    var App = (function (_super) {
        __extends(App, _super);
        function App() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        App.prototype.initialize = function (options) {
            this.uiApplication = 'sticker_creator';
            this.uiContextName = 'app';
            _super.prototype.initialize.call(this, options);
            this.addToDOM();
        };
        return App;
    }(_UiCore_1.UiContext));
}
main();


/***/ }),
```

and then:

<img width="603" alt="screen shot 2017-01-31 at 7 20 27 pm" src="https://cloud.githubusercontent.com/assets/24397438/22494205/679053ae-e7ea-11e6-9f0a-7bff97d1cb99.png">

To make matters worse, you can see from my example that the ES2015 output of `tsc` does not include `use strict`, but when we set `"module":"commonjs"` then tsc _does_ output `use strict`, which seems to be the opposite behavior of what we really need. This is the `commonjs` output of the above typescript module, with default setting for `noImplicitUseStrict`:

```js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var _UiCore_1 = require("@UiCore");
function main() {
    // Yay, we can use ES6 classes!
    var App = (function (_super) {
        __extends(App, _super);
        function App() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        App.prototype.initialize = function (options) {
            this.uiApplication = 'sticker_creator';
            this.uiContextName = 'app';
            _super.prototype.initialize.call(this, options);
            this.addToDOM();
        };
        return App;
    }(_UiCore_1.UiContext));
}
main();
```

This seems to be an inconsistency that should be fixed?

**Comment by trusktr** [comment]
I just noticed that TypeScript, when we've set `"module":"amd"`, will stick the `use strict` inside of the `define()` module body, and will place the `__extends` function outside of it, so it works:

```js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
define(["require", "exports", "@Core"], function (require, exports, _Core_1) {
    "use strict";
    function main() {
        // Yay, we can use ES6 classes!
        var App = (function (_super) {
            __extends(App, _super);
            function App() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            App.prototype.initialize = function (options) {
                _super.prototype.initialize.call(this, options);
                this.blah();
            };
            return App;
        }(_Core_1.BaseApp));
    }
    main();
});
```

I was wondering if maybe the problem was with `Core.BaseApp.prototype`, but apparently I can modify the prototype at runtime without any errors.

**Comment by trusktr** [comment]
@RyanCavanaugh This needs a fix. It is simply annoying to stick some code in some build system and for it to fail because it is in strict mode.

**Comment by mhegazy** [comment]
@trusktr please file a new issue and provide a self contained repro for the issue you are running into.

**Comment by trusktr** [comment]
@mhegazy Turns out the reason this happened to me is because I have a `BaseClass` that has a static method `extend` (similar to `BackBone`) that generates ES5-style classes, like

```js
var Foo = BaseClass.extend('Foo', {
  someMethod() {},
  // etc
})

// `BaseClass.extend` was assigned onto `Foo` (similar to ES6 static inheritance).
var Bar = Foo.extend('Bar', {...})
```

 Inside of `BaseClass.extend` the `constructor` property is defined like

```js
NewClass.prototype = Object.create(this.prototype, {
  constructor: {value: NewClass}
})
```

which means that by default the `constructor` property is not writable due to the `Object.create` call not having the `writable:true` property.

Then, when we use BaseClass or any class derived from it in TypeScript,

```ts
class Baz extends Bar {...}
```

then we experience the error in strict mode because `constructor` isn't writable by the `__extends` helper. In non-strict mode, the `__extends` helper does not throw an error, but the `constructor` assignment still fails (silently).

However, it's interesting to note that (in Chrome) we can use `class` syntax to extend a class who's `prototype.constructor` property is not writable, and the new ES6 class' `prototype.constructor` property will be the new class and will be writable:

```js
function Foo() {}

function Bar() {}
Bar.prototype = Object.create(Foo.prototype, {
  constructor: {value: Bar} // writable:false
})

function Baz() {}
Baz.prototype = Object.create(Bar.prototype)
~function() {
  "use strict"
  Baz.prototype.constructor = Baz // throws Error
}()
Baz.prototype.constructor = Baz // fails silently

class Lorem extends Bar {}
~function() {
  "use strict"
  Lorem.prototype.constructor = "anything" // no error in ES6 environment, only error in TypeScript.
}()
```

TLDR, it might still make sense to use `Object.defineProperty` because that seems to be similar to what ES6 classes do behind the scenes, so in doing so the TypeScript functionality will be more aligned with ES.

**Comment by trusktr** [comment]
Made a new issue here: https://github.com/Microsoft/TypeScript/issues/14069
